File: SearchInRotatedSortedArrayII.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        if not nums:
            return False
        l, r = 0, len(nums)-1
        while l < r:
            m = (l+r)//2
            if nums[m] == target:
                return True
            if nums[m] > nums[r]:
                # compare nums[m] and nums[r]
                if nums[m] > target and nums[l] <= target:
                    r = m
                else:
                    l = m + 1
            elif nums[m] < nums[r]:
                if nums[m] < target and nums[r] >= target:
                    l = m + 1
                else:
                    r = m
            else:
                r -= 1
        return True if nums[l] == target else False

nums = []
target = 2
print Solution().search(nums, target)

File: cpp/Search in Rotated Sorted Array II.cpp
class Solution {
public:
	bool search(int A[], int n, int target) {
		int first = 0, last = n;
		while (first != last) {
			const int mid = (first + last) >> 1;
			if (A[mid] == target) 
				return true;
			if (A[first] < A[mid]) {
				if (A[first] <= target && target < A[mid])
					last = mid;
				else
					first = mid + 1;
			} else if (A[first] > A[mid]) {
				if (A[mid] < target && target <= A[last - 1])
					first = mid + 1;
				else
					last = mid;
			} else {
				first ++;
			}
		}
		return false;
	}
};